### DISCUSSION TITLE
Field with control

### DISCUSSION THREAD

**Discussion by danyhiol** [vazco/uniforms#945] [Q&A] [answered]
What is the best way to wrap fields in a Col using SimpleSchema2 and the uniforms property?
Right now the form is displayed in rows of full-width column layout. I want to control the width of each column. I manage to make `AutoFields` display its content as a `Row` using the `element` property. Now I want to use the `uniforms` property in SimpleSchema2 to control the width of the field.
my first try was to create a custom field that just wraps the field in a Col.

```
function CustomCol(props) {
  ...
     return (<Col className="gutter-row" {...props.layout}>
        <AutoField name="" />
      </Col>);
}
export const CustomColField = connectField(CustomCol);
export default CustomColField;
```
Usage: 

```
firstName: {
    type: String,
    uniforms: {
      component: CustomColField,
      layout: { xs: 24, sm: 12, md: 6, lg: 6, xl: 6, xxl: 6 },
    },
  },
```
As a result, the UI is breaking. The form keeps re-rendering. Is there something wrong with my approach? Or is there a better way to achieve this?
I am using antd in the UI.

**Comment by radekmie** [comment, marked as answer]
Hi @danyhiol. I think it may be a perfect case for [overriding `AutoField`](https://uniforms.tools/docs/uth-autofield-algorithm#overriding-autofield). You could create a custom `AutoField`, that wouldn't return the original AntD components, but rather wrap them in a higher-order component rendering the wrapping component.

**Reply by danyhiol** [reply]
Ohhh, thank you for that. I have a few other questions:
~~1. Is there a way to replace the form AutoField with my CustomColField?~~ `autoFiel` in AutoForm or AutoFields
2. Can you advise me on how to implement a search on that using apollo and Graphql? The idea is to have a search field, take the user input, issue a get to the server, merge the result back to the fields options. It can be String or Array.

From your proposition above, can you provide a simple example using the `firstName` - `key` above?

This is what I tried so far:
```

function withLayout(
  props: GuaranteedProps<unknown> & Record<string, any>,
  uniforms: Context<unknown>,
) {
  const { label, name, field } = props;
  const { layout = { span: 24} } = field.uniforms;
  const LABEL = <FormattedMessage id={`app.label.${name}`} />;

  return (<Col className="gutter-row" {...layout}>
      {AutoField.defaultComponentDetector({ ...props, label: LABEL }, uniforms)}
  </Col>);
}

const AutoFieldProvider = ({ children }) => (
  <AutoField.componentDetectorContext.Provider value={withLayout}>
    {children}
  </AutoField.componentDetectorContext.Provider>
);

export default AutoFieldProvider;
```
The problem here is that nothing is being rendered. It would be nice to have some help here.
Thanks.

**Reply by danyhiol** [reply]
@radekmie any help here would be welcome. I think I am missing something in `{AutoField.defaultComponentDetector({ ...props, label: LABEL }, uniforms)}`

**Reply by radekmie** [reply]
It's because `withLayout` should return a component, not render it. I haven't tried it, but something like this should work:
```ts
import memoize from 'lodash/memoize';

function withLayoutImpl(Compoent) {
  return props => {
    const { field: { layout = { span: 24 } }, label, name } = props;
    return (
      <Col className="gutter-row" {...layout}>
        <Component {...props} label={<FormattedMessage id={`app.label.${name}`} />} /> 
      </Col>
    );
  };
}

const withLayout = memoize(withLayoutImpl);

function withLayoutWrapper(
  props: GuaranteedProps<unknown> & Record<string, any>,
  uniforms: Context<unknown>,
) {
  return withLayout(AutoField.defaultComponentDetector(props, uniforms));
}

const AutoFieldProvider = ({ children }) => (
  <AutoField.componentDetectorContext.Provider value={withLayoutWrapper}>
    {children}
  </AutoField.componentDetectorContext.Provider>
);

export default AutoFieldProvider;
```

**Reply by danyhiol** [reply]
After testing, this is not working. `props` from `withLayoutImpl` return `{ name: "", label: null}`. So `const { field: { layout = { span: 24 } }, label, name } = props;` is failing.  Doing `withLayout(AutoField.defaultComponentDetector(props, uniforms), props1);` and passing `<Component {...props}  {...props1} label={<FormattedMessage id={`app.label.${name}`} />} /> ` also is not working. Same problem as before, the component is empty on the ui.

**Reply by radekmie** [reply]
OK, I get it. We'll look into it soon.

**Reply by danyhiol** [reply]
Any news on this? Or a simple way to wrap fields in columns? I personally think it would be a big+ to do that because we rarely define forms with full width (at least on big forms). So many other devs may have this same issue.

**Reply by radekmie** [reply]
Hi @danyhiol, sorry for the delay. I've prepared [this demo](https://codesandbox.io/s/sleepy-torvalds-wmpod?file=/App.tsx) that bases on the idea I've posted above. You can safely use hooks in the component returned from `withLayoutImpl`, including `useField`.

We'll think about making a better/easier API for that in the future.

**Reply by danyhiol** [reply]
Hi @radekmie, after testing, this is indeed working as long as the schema doesn't contain the `component: SomeComponent` field. Once this field is added to the schema, the component is rendered without passing through the provider. Thus not applying the layout.
```
uniforms: {
  component: LongTextField,
  ...
}
```

**Reply by radekmie** [reply]
Yep, that's expected. However, if you already have a custom `AutoField`, you can stop using `component` in favor of your own field, e.g., `componentName` (or anything else), and derive the component based on that.

**Reply by danyhiol** [reply]
@radekmie fantastic
```
uniforms: {
      componentName: SelectField,
      layout: { xs: 24, md: 24 },
},
```
```
function withLayoutWrapper(
  props: GuaranteedProps<unknown> & Record<string, any>,
  uniforms: Context<unknown>,
) {
  if (props.componentName) {
    const DataFieldWrapper: React.FC = () => (
      <Col className="gutter-row" {...props.layout}>
        {React.createElement(props.componentName, props) }
      </Col>
    );

    return DataFieldWrapper;
  }
  return withLayout(AutoField.defaultComponentDetector(props, uniforms));
}
```
